### ISSUE TITLE
AttributeError: module 'warprnnt_pytorch.warp_rnnt' has no attribute 'gpu_rnnt'

### ISSUE THREAD

**Issue by ily6** [espnet/espnet#3948]
cuda: 11.3,  pytorch: 1.10.1, python: 3.8.5
I trained a conformer transducer model using aishell recipe, but got this as follows, could anyone give me a hand?
2022-01-08 19:48:09,317 (asr_train:563) WARNING: Skip DEBUG/INFO messages
2022-01-08 19:48:17,344 (asr:493) WARNING: num. model params: 32,575,561 (num. trained: 32,575,561 (100.0%))
2022-01-08 19:48:23,164 (io_utils:57) WARNING: [Experimental feature] Some preprocessing will be done for the mini-batch creation using Transformation(
    0: TimeWarp(max_time_warp=5, inplace=True, mode=PIL)
    1: FreqMask(F=30, n_mask=2, replace_with_zero=False, inplace=True)
    2: TimeMask(T=40, n_mask=2, replace_with_zero=False, inplace=True))
2022-01-08 19:48:23,167 (io_utils:57) WARNING: [Experimental feature] Some preprocessing will be done for the mini-batch creation using Transformation(
    0: TimeWarp(max_time_warp=5, inplace=True, mode=PIL)
    1: FreqMask(F=30, n_mask=2, replace_with_zero=False, inplace=True)
    2: TimeMask(T=40, n_mask=2, replace_with_zero=False, inplace=True))
^[[JException in main training loop: module 'warprnnt_pytorch.warp_rnnt' has no attribute 'gpu_rnnt'
Traceback (most recent call last):
  File "/home/lli/anaconda3/lib/python3.8/site-packages/chainer/training/trainer.py", line 316, in run
    update()
  File "/mnt/data3/lli/espnet/espnet/asr/pytorch_backend/asr.py", line 231, in update
    self.update_core()
  File "/mnt/data3/lli/espnet/espnet/asr/pytorch_backend/asr.py", line 195, in update_core
    data_parallel(self.model, x, range(self.ngpu)).mean() / self.accum_grad
  File "/home/lli/anaconda3/lib/python3.8/site-packages/torch/nn/parallel/data_parallel.py", line 228, in data_parallel
    return module(*inputs[0], **module_kwargs[0])
  File "/home/lli/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/data3/lli/espnet/espnet/nets/pytorch_backend/e2e_asr_transducer.py", line 418, in forward
    losses = self.transducer_tasks(
  File "/home/lli/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/data3/lli/espnet/espnet/nets/pytorch_backend/transducer/transducer_tasks.py", line 440, in forward
    joint_out, trans_loss = self.compute_transducer_loss(
  File "/mnt/data3/lli/espnet/espnet/nets/pytorch_backend/transducer/transducer_tasks.py", line 170, in compute_transducer_loss
    loss_trans = self.transducer_loss(joint_out, target, t_len, u_len)
  File "/home/lli/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/data3/lli/espnet/tools/warp-transducer/pytorch_binding/warprnnt_pytorch/__init__.py", line 104, in forward
    return self.loss(acts, labels, act_lens, label_lens, self.blank, self.reduction, self.fastemit_lambda)
  File "/mnt/data3/lli/espnet/tools/warp-transducer/pytorch_binding/warprnnt_pytorch/__init__.py", line 24, in forward
    loss_func = warp_rnnt.gpu_rnnt if is_cuda else warp_rnnt.cpu_rnnt
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "/mnt/data3/lli/espnet/egs/aishell/asr1/../../../espnet/bin/asr_train.py", line 639, in <module>
    main(sys.argv[1:])
  File "/mnt/data3/lli/espnet/egs/aishell/asr1/../../../espnet/bin/asr_train.py", line 625, in main
    train(args)
  File "/mnt/data3/lli/espnet/espnet/asr/pytorch_backend/asr.py", line 965, in train
    trainer.run()
 File "/home/lli/anaconda3/lib/python3.8/site-packages/chainer/training/trainer.py", line 349, in run
    six.reraise(*exc_info)
  File "/home/lli/anaconda3/lib/python3.8/site-packages/six.py", line 703, in reraise
    raise value
  File "/home/lli/anaconda3/lib/python3.8/site-packages/chainer/training/trainer.py", line 316, in run
    update()
  File "/mnt/data3/lli/espnet/espnet/asr/pytorch_backend/asr.py", line 231, in update
    self.update_core()
  File "/mnt/data3/lli/espnet/espnet/asr/pytorch_backend/asr.py", line 195, in update_core
    data_parallel(self.model, x, range(self.ngpu)).mean() / self.accum_grad
  File "/home/lli/anaconda3/lib/python3.8/site-packages/torch/nn/parallel/data_parallel.py", line 228, in data_parallel
    return module(*inputs[0], **module_kwargs[0])
  File "/home/lli/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/data3/lli/espnet/espnet/nets/pytorch_backend/e2e_asr_transducer.py", line 418, in forward
    losses = self.transducer_tasks(
  File "/home/lli/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/data3/lli/espnet/espnet/nets/pytorch_backend/transducer/transducer_tasks.py", line 440, in forward
    joint_out, trans_loss = self.compute_transducer_loss(
  File "/mnt/data3/lli/espnet/espnet/nets/pytorch_backend/transducer/transducer_tasks.py", line 170, in compute_transducer_loss
    loss_trans = self.transducer_loss(joint_out, target, t_len, u_len)
  File "/home/lli/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/data3/lli/espnet/tools/warp-transducer/pytorch_binding/warprnnt_pytorch/__init__.py", line 104, in forward
    return self.loss(acts, labels, act_lens, label_lens, self.blank, self.reduction, self.fastemit_lambda)
  File "/mnt/data3/lli/espnet/tools/warp-transducer/pytorch_binding/warprnnt_pytorch/__init__.py", line 24, in forward
    loss_func = warp_rnnt.gpu_rnnt if is_cuda else warp_rnnt.cpu_rnnt
AttributeError: module 'warprnnt_pytorch.warp_rnnt' has no attribute 'gpu_rnnt'

**Comment by b-flo** [comment]
Hello,

I'm not sure about this one to be honest. From what I gather, it could be caused by a wrong/missing CUDA related env variable. Or something went wrong during install.
Could you paste your `CPATH `and `LD_LIBRARY_PATH `please?

**Comment by ily6** [comment]
Hello, I found that CPATH and LD_LIBRARY_PATH are empty.

**Comment by csukuangfj** [comment]
One useful information for debugging is to print
```python3
print(warprnnt_pytorch.warp_rnnt.__file__)
print(dir(warprnnt_pytorch.warp_rnnt))
```

**Comment by ily6** [comment]
Thanks, I have another question. I still trained the conformer transducer model using aishell recipe, but got this:
ValueError:  noam-adim option should be set to use Noam scheduler.

**Comment by b-flo** [comment]
You need to set `noam-adim` in your training config if you are using `opt: noam`, see [this example](https://github.com/espnet/espnet/blob/7112ef7ac32dd1b2ee98cca6a268b0ff0ae96f5c/egs/voxforge/asr1/conf/tuning/transducer/train_conformer-rnn_transducer.yaml).
The value should be the most dominant attention dimension (`d_hidden`) in your architecture but any value can be used.

**Comment by ily6** [comment]
Ok, thanks very much.

**Comment by ily6** [comment]
The results of conformer transducer using aishell recipe are as follows:
	|  SPKR        |   # Snt       # Wrd   |   Corr       Sub        Del       Ins        Err     S.Err      |
	|  Sum/Avg  |  14326     205341   |    1.0      21.9       77.1       0.1       99.1     100.0   |dev
	|  Sum/Avg  |  7176      104765    |   0.9       21.4       77.7       0.1      99.2      100.0   |test
And the loss:
![图片1](https://user-images.githubusercontent.com/97342412/150530424-4d2ff0c4-5304-44c1-ae83-47225680fe01.png)

**Comment by b-flo** [comment]
Interesting, someone recently reported the same behaviour with a Transducer system and Aishell-1.

Could you share the run script + logs (and the configs if there are different from the original ones) please?
@yusshino Did you encounter a similar issue in your experiments?

**Comment by ily6** [comment]
train_conformer-rnn_transducer.yaml
# minibatch related
batch-size: 16
maxlen-in: 512
maxlen-out: 150

# optimization related
criterion: loss
early-stop-criterion: "validation/main/loss"
sortagrad: 0
opt: noam
noam-adim: 256
transformer-lr: 1.0
transformer-warmup-steps: 25000
epochs: 50
patience: 0
accum-grad: 2
grad-clip: 5.0

# network architecture
## general
custom-enc-positional-encoding-type: rel_pos
custom-enc-self-attn-type: rel_self_attn
custom-enc-pw-activation-type: swish
## encoder related
etype: custom
custom-enc-input-layer: vgg2l
enc-block-arch:
        - type: conformer
          d_hidden: 512
          d_ff: 2048
          heads: 4
          macaron_style: True
          use_conv_mod: True
          conv_mod_kernel: 15
          dropout-rate: 0.3
          att-dropout-rate: 0.3
enc-block-repeat: 12
## decoder related
dtype: lstm
dlayers: 1
dec-embed-dim: 1024
dunits: 512
dropout-rate-embed-decoder: 0.2
dropout-rate-decoder: 0.1
## joint network related
joint-dim: 512

# transducer related
model-module: "espnet.nets.pytorch_backend.e2e_asr_transducer:E2E"

# reporter related
report-wer: True
report-cer: True
[result.txt](https://github.com/espnet/espnet/files/7914081/result.txt)
[train.log](https://github.com/espnet/espnet/files/7914085/train.log)

run.sh
#!/usr/bin/env bash

# Copyright 2017 Johns Hopkins University (Shinji Watanabe)
#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)

. ./path.sh || exit 1;
. ./cmd.sh || exit 1;

# general configuration
backend=pytorch
stage=0        # start from 0 if you need to start from data preparation
stop_stage=100
ngpu=1         # number of gpus ("0" uses cpu, otherwise use gpu)
debugmode=1
dumpdir=dump   # directory to dump full features
N=0            # number of minibatches to be used (mainly for debugging). "0" uses all minibatches.
verbose=0      # verbose option
resume=        # Resume the training from snapshot

# feature configuration
do_delta=false

preprocess_config=conf/specaug.yaml
train_config=conf/tuning/transducer/train_conformer-rnn_transducer.yaml
lm_config=-
decode_config=conf/tuning/transducer/decode_default.yaml

# rnnlm related
lm_resume=         # specify a snapshot file to resume LM training
lmtag=             # tag for managing LMs

# ngram
ngramtag=
n_gram=4

# decoding parameter
recog_model=model.acc.best # set a model to be used for decoding: 'model.acc.best' or 'model.loss.best'
n_average=10

# data
data=/Share/lli/data/aishell
data_url=www.openslr.org/resources/33

# exp tag
tag="" # tag for managing experiments.

. utils/parse_options.sh || exit 1;

# Set bash to 'debug' mode, it will exit on :
# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
set -e
set -u
set -o pipefail

train_set=train_sp
train_dev=dev
recog_set="dev test"

if [ ${stage} -le -1 ] && [ ${stop_stage} -ge -1 ]; then
    echo "stage -1: Data Download"
    local/download_and_untar.sh ${data} ${data_url} data_aishell
    local/download_and_untar.sh ${data} ${data_url} resource_aishell
fi

if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
    ### Task dependent. You have to make data the following preparation part by yourself.
    ### But you can utilize Kaldi recipes in most cases
    echo "stage 0: Data preparation"
    local/aishell_data_prep.sh ${data}/data_aishell/wav ${data}/data_aishell/transcript
    # remove space in text
    for x in train dev test; do
        cp data/${x}/text data/${x}/text.org
        paste -d " " <(cut -f 1 -d" " data/${x}/text.org) <(cut -f 2- -d" " data/${x}/text.org | tr -d " ") \
            > data/${x}/text
        rm data/${x}/text.org
    done
fi

feat_tr_dir=${dumpdir}/${train_set}/delta${do_delta}; mkdir -p ${feat_tr_dir}
feat_dt_dir=${dumpdir}/${train_dev}/delta${do_delta}; mkdir -p ${feat_dt_dir}
if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
    ### Task dependent. You have to design training and dev sets by yourself.
    ### But you can utilize Kaldi recipes in most cases
    echo "stage 1: Feature Generation"
    fbankdir=fbank
    # Generate the fbank features; by default 80-dimensional fbanks with pitch on each frame
    steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 32 --write_utt2num_frames true \
        data/train exp/make_fbank/train ${fbankdir}
    utils/fix_data_dir.sh data/train
    steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 10 --write_utt2num_frames true \
        data/dev exp/make_fbank/dev ${fbankdir}
    utils/fix_data_dir.sh data/dev
    steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 10 --write_utt2num_frames true \
        data/test exp/make_fbank/test ${fbankdir}
    utils/fix_data_dir.sh data/test

    # speed-perturbed
    utils/perturb_data_dir_speed.sh 0.9 data/train data/temp1
    utils/perturb_data_dir_speed.sh 1.0 data/train data/temp2
    utils/perturb_data_dir_speed.sh 1.1 data/train data/temp3
    utils/combine_data.sh --extra-files utt2uniq data/${train_set} data/temp1 data/temp2 data/temp3
    rm -r data/temp1 data/temp2 data/temp3
    steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 32 --write_utt2num_frames true \
        data/${train_set} exp/make_fbank/${train_set} ${fbankdir}
    utils/fix_data_dir.sh data/${train_set}

    # compute global CMVN
    compute-cmvn-stats scp:data/${train_set}/feats.scp data/${train_set}/cmvn.ark

    # dump features for training
    split_dir=$(echo $PWD | awk -F "/" '{print $NF "/" $(NF-1)}')
    if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d ${feat_tr_dir}/storage ]; then
    utils/create_split_dir.pl \
        /export/a{11,12,13,14}/${USER}/espnet-data/egs/${split_dir}/dump/${train_set}/delta${do_delta}/storage \
        ${feat_tr_dir}/storage
    fi
    if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d ${feat_dt_dir}/storage ]; then
    utils/create_split_dir.pl \
        /export/a{11,12,13,14}/${USER}/espnet-data/egs/${split_dir}/dump/${train_dev}/delta${do_delta}/storage \
        ${feat_dt_dir}/storage
    fi
    dump.sh --cmd "$train_cmd" --nj 32 --do_delta ${do_delta} \
        data/${train_set}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/train ${feat_tr_dir}
    for rtask in ${recog_set}; do
        feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}; mkdir -p ${feat_recog_dir}
        dump.sh --cmd "$train_cmd" --nj 10 --do_delta ${do_delta} \
            data/${rtask}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/recog/${rtask} \
            ${feat_recog_dir}
    done
fi

dict=data/lang_1char/${train_set}_units.txt
echo "dictionary: ${dict}"
if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
    ### Task dependent. You have to check non-linguistic symbols used in the corpus.
    echo "stage 2: Dictionary and Json Data Preparation"
    mkdir -p data/lang_1char/

    echo "make a dictionary"
    echo "<unk> 1" > ${dict} # <unk> must be 1, 0 will be used for "blank" in CTC
    text2token.py -s 1 -n 1 data/${train_set}/text | cut -f 2- -d" " | tr " " "\n" \
    | sort | uniq | grep -v -e '^\s*$' | awk '{print $0 " " NR+1}' >> ${dict}
    wc -l ${dict}

    echo "make json files"
    data2json.sh --feat ${feat_tr_dir}/feats.scp \
		 data/${train_set} ${dict} > ${feat_tr_dir}/data.json
    for rtask in ${recog_set}; do
        feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}
        data2json.sh --feat ${feat_recog_dir}/feats.scp \
		     data/${rtask} ${dict} > ${feat_recog_dir}/data.json
    done
fi

# you can skip this and remove --rnnlm option in the recognition (stage 5)
#if [ -z ${lmtag} ]; then
    #lmtag=$(basename ${lm_config%.*})
#fi
lmexpname=train_rnnlm_${backend}_${lmtag}
lmexpdir=exp/${lmexpname}
mkdir -p ${lmexpdir}

ngramexpname=train_ngram
ngramexpdir=exp/${ngramexpname}
if [ -z ${ngramtag} ]; then
    ngramtag=${n_gram}
fi
mkdir -p ${ngramexpdir}

if [ ${stage} -le -1 ] && [ ${stop_stage} -ge 3 ]; then
    echo "stage 3: LM Preparation"
    lmdatadir=data/local/lm_train
    mkdir -p ${lmdatadir}
    text2token.py -s 1 -n 1 data/train/text | cut -f 2- -d" " \
        > ${lmdatadir}/train.txt
    text2token.py -s 1 -n 1 data/${train_dev}/text | cut -f 2- -d" " \
        > ${lmdatadir}/valid.txt

    ${cuda_cmd} --gpu ${ngpu} ${lmexpdir}/train.log \
        lm_train.py \
        --config ${lm_config} \
        --ngpu ${ngpu} \
        --backend ${backend} \
        --verbose 1 \
        --outdir ${lmexpdir} \
        --tensorboard-dir tensorboard/${lmexpname} \
        --train-label ${lmdatadir}/train.txt \
        --valid-label ${lmdatadir}/valid.txt \
        --resume ${lm_resume} \
        --dict ${dict}
    
    lmplz --discount_fallback -o ${n_gram} <${lmdatadir}/train.txt > ${ngramexpdir}/${n_gram}gram.arpa
    build_binary -s ${ngramexpdir}/${n_gram}gram.arpa ${ngramexpdir}/${n_gram}gram.bin
fi


if [ -z ${tag} ]; then
    expname=${train_set}_${backend}_$(basename ${train_config%.*})
    if ${do_delta}; then
        expname=${expname}_delta
    fi
    if [ -n "${preprocess_config}" ]; then
        expname=${expname}_$(basename ${preprocess_config%.*})
    fi
else
    expname=${train_set}_${backend}_${tag}
fi
expdir=exp/${expname}
mkdir -p ${expdir}

if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ]; then
    echo "stage 4: Network Training"
    ${cuda_cmd} --gpu ${ngpu} ${expdir}/train.log \
        asr_train.py \
        --config ${train_config} \
        --preprocess-conf ${preprocess_config} \
        --ngpu ${ngpu} \
        --backend ${backend} \
        --outdir ${expdir}/results \
        --tensorboard-dir tensorboard/${expname} \
        --debugmode ${debugmode} \
        --dict ${dict} \
        --debugdir ${expdir} \
        --minibatches ${N} \
        --verbose ${verbose} \
        --resume ${resume} \
        --train-json ${feat_tr_dir}/data.json \
        --valid-json ${feat_dt_dir}/data.json
fi

if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ]; then
    echo "stage 5: Decoding"
    nj=32
    if [[ $(get_yaml.py ${train_config} model-module) = *transformer* ]] || \
           [[ $(get_yaml.py ${train_config} model-module) = *conformer* ]] || \
           [[ $(get_yaml.py ${train_config} etype) = custom ]] || \
           [[ $(get_yaml.py ${train_config} dtype) = custom ]]; then
        recog_model=model.last${n_average}.avg.best
        average_checkpoints.py --backend ${backend} \
        		       --snapshots ${expdir}/results/snapshot.ep.* \
        		       --out ${expdir}/results/${recog_model} \
        		       --num ${n_average}
    fi

    if [[ $(get_yaml.py ${train_config} model-module) = *transducer* ]]; then
        echo "[info]: transducer model does not support '--api v2'" \
             "(hence ngram is ignored)"
        recog_v2_opts=""
    else
        recog_v2_opts="--ngram-model ${ngramexpdir}/${n_gram}gram.bin --api v2"
    fi

    pids=() # initialize pids
    for rtask in ${recog_set}; do
    (
        decode_dir=decode_${rtask}_$(basename ${decode_config%.*})_${lmtag}_${ngramtag}
        feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}

        # split data
        splitjson.py --parts ${nj} ${feat_recog_dir}/data.json

        #### use CPU for decoding
        ngpu=0

        ${decode_cmd} JOB=1:${nj} ${expdir}/${decode_dir}/log/decode.JOB.log \
            asr_recog.py \
            --config ${decode_config} \
            --ngpu ${ngpu} \
            --backend ${backend} \
            --batchsize 0 \
            --recog-json ${feat_recog_dir}/split${nj}utt/data.JOB.json \
            --result-label ${expdir}/${decode_dir}/data.JOB.json \
            --model ${expdir}/results/${recog_model}  \
            #--rnnlm ${lmexpdir}/rnnlm.model.best \
            ${recog_v2_opts}

        score_sclite.sh ${expdir}/${decode_dir} ${dict}

    ) &
    pids+=($!) # store background pids
    done
    i=0; for pid in "${pids[@]}"; do wait ${pid} || ((++i)); done
    [ ${i} -gt 0 ] && echo "$0: ${i} background jobs are failed." && false
    echo "Finished"
fi

**Comment by yusshino** [comment]
Thanks a lot for your contact. Umm, interesting..

I've noticed two minor things (listed below) that's different from my usual setup, but I don't think they damage CERs so drastically (like CER=99.1%). 

I don't know why the training didn't go well at the moment, but I'll post here if I find something suspicious. 

- batch-size
  - I usually use `batch-size=64` with `accum-grad=2` when I use 1 GPU (i.e. effective batch size = 64 * 2 * 1)
  - Using configurations like (batch-size=16, accum-grad=8, ngpu=1) or (batch-size=32, accum-grad=2, ngpu=2) or whatever to make effective batch size of 128 might reproduce our AISHELL-1 numbers in [ASRU 2021 paper](https://arxiv.org/abs/2201.05420). 
- noam-adim
  - `noam-adim=512` might improve the performance (If I remember correctly, it is recommended to match with `d_hidden`)

**Comment by yusshino** [comment]
I've reproduced the issue (got CER over 99%). 

Strangely, when I've tried with `n_average=1` (instead of 10), I didn't see the issue (got CER 5.7%/6.3% on dev/test). I've also tried `n_average=2` and `n_average=3` and observed the issue again (CERs over 99%). There might be some issue related to model averaging.

FYI, I've used `train_conformer-rnn_transducer.yaml` with following modifications to match with the new parameter names. I've used `v.0.10.5` (I believe I've never had such issue when I've tested with `v.0.10.3` a few months ago). 

```
-transformer-lr: 1.0
-transformer-warmup-steps: 25000
+noam-lr: 1.0
+noam-adim: 512
+optimizer-warmup-steps: 25000
```

**Comment by b-flo** [comment]
Interesting, thank you very much @yusshino ! I'll investigate the issue, I'm not sure other ASR models ever suffered from model averaging.

Btw, what about performance with current version? Is that on par with previous ones?

**Comment by yusshino** [comment]
Thanks for comment! 

> Btw, what about performance with current version? Is that on par with previous ones?

Above numbers (CER 5.7%/6.3%) are worse than previous numbers (I've got 4.5%/5.1% with `v.0.10.3`, almost identical with ASRU numbers), but they were obtained at epoch 53 (out of 100) with `n_average=1` (instead of 10), so I believe they are not directly comparable with previous numbers.

**Comment by b-flo** [comment]
Hum, it seems a bit high compared to some of our ASRU experiments (cf. gdoc) but I also found model averaging to be quite important for performance . Outside of fixes, ESPnet1 Transducer won't be further modified so I wan't to be sure it's in proper shape.

@ily6 Could you report your results after tweaking model averaging please? I'll also try to train the model with one GPU based on @yusshino info.

**Comment by yusshino** [comment]
Yeah, the CER might be a bit high. I'll report numbers when I reach epoch 100. Also I'm thinking to evaluate the v.0.10.3 model with `n_average=1` for fair comparison.

**Comment by b-flo** [comment]
Oh, thanks a lot! I'll wait for your report before allocating ressources on that then!

**Comment by ily6** [comment]
Thank you both so much. I've tried with n_average=1, and the results:
|   SPKR         |  # Snt      # Wrd    |  Corr       Sub      Del      Ins       Err      S.Err  |
|   Sum/Avg  |  7176      104765   |  93.9       5.8       0.3       0.1       6.2      42.4   |test
|  Sum/Avg   |  14326     205341  |  94.7       5.1       0.2       0.1       5.5      38.7   |dev

**Comment by yusshino** [comment]
Thanks for sharing it. Nice results! 
Umm, I still don't know why `n_average` affects the performance so much..

**Comment by b-flo** [comment]
Thanks a lot for the report @ily6 ! And yeah, Err is high without averaging... It could be interesting point for further optimization.

I quickly checked if related modules were recently changed but nothing suspicious at first glance.
Btw @yusshino which torch version are you using? I just recalled being warned of some deprecated operation during model averaging, not sure if I was notified of a different behaviour though.

**Comment by yusshino** [comment]
I've used PyTorch 1.6.0 for both `v.0.10.5` and `v.0.10.3` experiments.

**Comment by yusshino** [comment]
I've compared `v.0.10.3` and `v.0.10.5` at epoch 80.
(I've needed to shutdown my machine for maintenance before reaching epoch 100, sorry..)

The result is as follows. Significant degradation in CER was observed. Something might not be working correctly for `v.0.10.5`.
```
v.0.10.3 (epoch 80, n-average 1): CER 4.9% (dev), 5.5% (test)
v.0.10.5 (epoch 80, n-average 1): CER 5.6% (dev), 6.2% (test)
```

By the way, I don't know if it's relevant but I've noticed that the number of parameters has significantly changed, although I've used exactly the same training config (train.yaml).
```
(v.0.10.3）
2021-10-29 20:00:57,422 (asr:503) WARNING: num. model params: 87,538,249 (num. trained: 87,538,249 (100.0%))

(v.0.10.5)
2022-01-21 18:05:17,839 (asr:499) WARNING: num. model params: 32,575,561 (num. trained: 32,575,561 (100.0%))
```

P.S.
Just for your information, here is some numbers from epoch 100:
```
v.0.10.3 (epoch 100, n-average 10): CER 4.5% (dev), 5.1% (test)
v.0.10.3 (epoch 100, n-average 1): CER 4.8% (dev), 5.4% (test)
```

**Comment by b-flo** [comment]
`87,538,249` parameters before vs. `32,575,561` now, this is quite a difference..! Could you print the model object info, please?
Do you also have the information for v.0.10.4?

Looking at the changes v.0.10.3 and v.0.10.5, I don't see any significant modification of transducer-related files except for #3697. I modified the custom encoder/decoder building and also some default parameters values in this one, it may be related.

**Comment by yusshino** [comment]
There was no diff in `print(model)` output, but when I do

```
for name, param in model.named_parameters():
    print(name, param.numel())
```

I found that the following part is missing in encoder.encoders.[1-11] in the `v.0.10.5` model. The total size of these missing ones is 54962688 (4996608 per layer), which matches with the difference between `87,538,249` and `32,575,561`. (only encoder.encoders.0 has this part. Note that I only paste 1st layer but same for 2nd-11th layers.)

```
encoder.encoders.1.feed_forward.w_1.weight 1048576
encoder.encoders.1.feed_forward.w_1.bias 2048
encoder.encoders.1.feed_forward.w_2.weight 1048576
encoder.encoders.1.feed_forward.w_2.bias 512
encoder.encoders.1.feed_forward_macaron.w_1.weight 1048576
encoder.encoders.1.feed_forward_macaron.w_1.bias 2048
encoder.encoders.1.feed_forward_macaron.w_2.weight 1048576
encoder.encoders.1.feed_forward_macaron.w_2.bias 512
encoder.encoders.1.conv_module.pointwise_conv1.weight 524288
encoder.encoders.1.conv_module.pointwise_conv1.bias 1024
encoder.encoders.1.conv_module.depthwise_conv.weight 7680
encoder.encoders.1.conv_module.depthwise_conv.bias 512
encoder.encoders.1.conv_module.norm.weight 512
encoder.encoders.1.conv_module.norm.bias 512
encoder.encoders.1.conv_module.pointwise_conv2.weight 262144
encoder.encoders.1.conv_module.pointwise_conv2.bias 512
```

It seems that somehow these ones are not found by `model.parameters()`.

**Comment by yusshino** [comment]
`v.0.10.4` seems to have the same issue.

```
2022-01-26 10:16:49,826 (asr:499) WARNING: num. model params: 32,575,561 (num. trained: 32,575,561 (100.0%))
```

**Comment by b-flo** [comment]
Interesting, #3697 should be the culprit then (if it's related to my changes).

Btw, unrelated to this issue but `custom-enc-input-dropout-rate` and `custom-enc-input-pos-enc-dropout-rate` were introduced in v.0.10.4. Previously, the value was set automatically according to the encoder architecture parameters.
Following your config, `custom-enc-input-dropout-rate: 0.3` should be added to the training config.

**Comment by yusshino** [comment]
Thanks for the information! I'll add the dropout setting when I use newer versions.

**Comment by b-flo** [comment]
Fixed in #4020. Closing the issue.
